In [2]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score #Confusion matrix 수업 때 진행할 예정 
from sklearn.model_selection import KFold
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score#임포트 
import numpy as np
import pandas as pd

In [3]:
fold_iris = load_iris()
features = fold_iris.data

label= fold_iris.target

In [4]:
features

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [5]:
label

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

## 1. Train/test/split

- 기본 모델로 나눠서 실제 과적합이 어떤 식으로 발생하는지?

In [37]:
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.2, random_state=111)

In [43]:
model = DecisionTreeClassifier()
model_sp1 = DecisionTreeClassifier(random_state=111, max_depth=3)
model.fit(X_train, y_train) #모델 학습
model_sp1.fit(X_train, y_train) #모델 학습
pred_train = model.predict(X_train) #train 예측
pred_test = model.predict(X_test) # test예측

pred_train_sp = model_sp1.predict(X_train) #train 예측
pred_test_sp = model_sp1.predict(X_test) # test예측

In [44]:
print(accuracy_score(y_train, pred_train))
print(accuracy_score(y_test, pred_test))

1.0
0.9


In [45]:
print(accuracy_score(y_train, pred_train_sp))
print(accuracy_score(y_test, pred_test_sp))

0.9916666666666667
0.9


- max_depth=3 으로 진행하면 기존 train 1에서 -> 0.9916666666666667 train 정확도가 줄었다.
- decisionTree -> 가지치기 형식으로 뻗어나는 구조 -> 데이터를 나눠가면서 분리하는데 max_depth 라는 것이 최대 깊이를 말하는데 이 부분을 최소화로 줄이니깐 (3) 으로 줄이니 -> train 정확도가 좀 줄어들었다.


- iris 데이터라서 너무 작기 때문에 그럴 수 있지만-> 동일한 원리로 보면 최대 깊이를 지정한다는 것은
- 최대 깊이를 지정하지 않았을 때 -> 데이터의 모든 것을 다 분리하고 학습
- 최대 깊이를 지정한 경우는 -> 분리하는 과정에서 3개 까지만 분리하고 멈춘 것 ( 아직 분리할 것들이 남았지만 그 부분은 학습하지 않음 )
- 기존에 1이아닌 정확도가 조금 줄게 되었고, 이부분이 과적합을 막기 위해서 하이퍼파라미터를 튜닝해서 과적합을 줄이고 일반화를 하기 위해 진행한 것

- train 1
- test 0.9

- 과적합으로 인해 문제가 발생했다.

## 교차검증을 통해서 과연 과적합이 나눠지는지?

In [46]:
from sklearn.model_selection import KFold

kfold= KFold(n_splits=5) #데이터셋을 분리하는 것

cv_accuracy_train=[]
cv_accuracy_test=[]
kf_model = DecisionTreeClassifier(random_state=111, max_depth=3)

In [47]:
kfold

KFold(n_splits=5, random_state=None, shuffle=False)

In [48]:
n_iter = 0 #5번 진행하니 초깃값 설정

for train_idx, test_idx in kfold.split(features):
    X_train, X_test = features[train_idx], features[test_idx]
    y_train, y_test = label[train_idx], label[test_idx]
    
    #학습을 위해서 dt배우기
    kf_model.fit(X_train, y_train)
    #예측
    kf_pred_train = kf_model.predict(X_train)
    df_pred_test = kf_model.predict(X_test)
    
    #정확도를 5번 측정하면서 값에 넣을 것이니
    n_iter +=1
    accuracy_train = np.round(accuracy_score(y_train, kf_pred_train),4)
    accuracy_test = np.round(accuracy_score(y_test, df_pred_test),4)
    
    print('\n{} 번 train 교차 검증 정확도 :{} ,test 교차 검증 정확도 :{} '.format(n_iter,accuracy_train, accuracy_test))
    
    cv_accuracy_train.append(accuracy_train)
    cv_accuracy_test.append(accuracy_test)
    
print('train 평균 정확도',np.mean(cv_accuracy_train))
print('test 평균 정확도',np.mean(cv_accuracy_test))


1 번 train 교차 검증 정확도 :0.975 ,test 교차 검증 정확도 :1.0 

2 번 train 교차 검증 정확도 :0.975 ,test 교차 검증 정확도 :0.9667 

3 번 train 교차 검증 정확도 :0.9917 ,test 교차 검증 정확도 :0.8333 

4 번 train 교차 검증 정확도 :0.9833 ,test 교차 검증 정확도 :0.9333 

5 번 train 교차 검증 정확도 :0.975 ,test 교차 검증 정확도 :0.7 
train 평균 정확도 0.9799999999999999
test 평균 정확도 0.88666


- train은 확실히 줄어들었다.
- 기존에 하이퍼파라미터가 없는 경우는 다 1
- 지금은 0.9 정도로 떨어짐
- 신기한 건 교차검증정확도가 1도 나온다? 

- train 0.975 test는 1이 나왔다?
- 이 부분 어떻게 해석해야 하는가?

### 이 부분을 해결하기 위해서는?

- 과적합을 막기 위해서 Skf 방식을 사용해 보자!

- 모델링에서 과적합을 막기위해 하이퍼파라미터를 추가할 수 있다.

In [28]:
label

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

## Skf 모델 진행

- kfold의 문제점은?
- 정답라벨이 0,1,2 총 3개 -> Kfold는 랜덤하게 데이터를 추출하기 때문에 정답 비중이 고르게 분포하지 않을 수 있다.
- 그러다보니 정답레이블이 불균형또는 다중분류인 경우는 SKF를 통해서 정답 라벨 불균형을 해소해 준다!

In [49]:
from sklearn.model_selection import StratifiedKFold

skf_iris = StratifiedKFold(n_splits=3)
cnt_iter=0

In [50]:
skf_iris

StratifiedKFold(n_splits=3, random_state=None, shuffle=False)

In [51]:
n_iter=0

skf_cv_accuracy_train = []
skf_cv_accuracy_test = []
skf_model = DecisionTreeClassifier(random_state=111, max_depth=3)

for train_idx, test_idx in skf_iris.split(features, label):
    X_train, X_test = features[train_idx], features[test_idx]
    y_train, y_test = label[train_idx], label[test_idx]
    
    #학습을 위해서 dt배우기
    skf_model.fit(X_train, y_train)
    #예측
    skf_pred_train = skf_model.predict(X_train)
    skf_pred_test = skf_model.predict(X_test)
    
    #정확도를 5번 측정하면서 값에 넣을 것이니
    n_iter +=1
    accuracy_train = np.round(accuracy_score(y_train, skf_pred_train),4)
    accuracy_test = np.round(accuracy_score(y_test, skf_pred_test),4)
    
    print('\n{} 번 train 교차 검증 정확도 :{} ,test 교차 검증 정확도 :{} '.format(n_iter,accuracy_train, accuracy_test))
    
    skf_cv_accuracy_train.append(accuracy_train)
    skf_cv_accuracy_test.append(accuracy_test)
    
print('train 평균 정확도',np.mean(skf_cv_accuracy_train))
print('test 평균 정확도',np.mean(skf_cv_accuracy_test))


1 번 train 교차 검증 정확도 :0.95 ,test 교차 검증 정확도 :0.98 

2 번 train 교차 검증 정확도 :0.99 ,test 교차 검증 정확도 :0.94 

3 번 train 교차 검증 정확도 :0.97 ,test 교차 검증 정확도 :0.96 
train 평균 정확도 0.9700000000000001
test 평균 정확도 0.96


### train에 대해서 과적합이 되는 것이 iris 작은 데이터라서 그런 것인지?
### 최소 타이타닉 데이터 정도가 되면 달라지는 것인가?

### 모델링에서 과적합을 막기위해 하이퍼파라미터를 추가할 수 있다.
- 이 부분을 추가하게 되면 과적합이 정말 줄어드는지 확인하기!

- kfold
- 학습시에 0,1,2 이 정답데이터가 고르게 분포 되어 있더라도, 랜덤하게 추출하다 보니 
- 0,1 만 있는데이터, 0만 잇는 데이터, 0,2 만있는 데이터 이런식으로 추출이 될 가능성이 높다. 
- 추출의 비중도 0(10개) 1(5개)  예를 들어
- 이런 식으로 학습할 데이터의 라벨이 달라지면 정확도 등의 문제가 올 수 있다.
- Skold 
- 학습시에 정답레이블의 불균형을 맞춰준다고 햇으니
- 0,1,2 라벨이 잇는데
- 0,1,2 (0(10),1(10),2(10)) 씩 비중도 맞춰줘서 학습하게 된다.

- 정답레이블에 대한 불균형은 항상 모델링을 진행할 때 꼭 체크해야 하는 부분입니다.
- iris 150개 정도 밖에 되지 않아서 다른 데이터 set으로 과제를 진행하자!

In [55]:
import pandas as pd
iris=load_iris()

iris_df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
iris_df['label'] = iris.target

In [56]:
iris_df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)   
0                  5.1               3.5                1.4               0.2  \
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     label  
0        0  
1        0  
2        0  
3        0  
4        0  
..     ...  
145      2  
146      2  
147      2  
148      2  
149      2  

[150 rows x 5 columns]

In [58]:
kfold= KFold(n_splits=5) 
n_iter = 0

for train_idx, test_idx in kfold.split(iris_df):
    n_iter +=1
    label_train = iris_df['label'].iloc[train_idx]
    label_test = iris_df['label'].iloc[test_idx]
    print('학습 정답 레이블', label_train.value_counts())
    print('학습 정답 레이블', label_test.value_counts())

학습 정답 레이블 label
1    50
2    50
0    20
Name: count, dtype: int64
학습 정답 레이블 label
0    30
Name: count, dtype: int64
학습 정답 레이블 label
2    50
1    40
0    30
Name: count, dtype: int64
학습 정답 레이블 label
0    20
1    10
Name: count, dtype: int64
학습 정답 레이블 label
0    50
2    50
1    20
Name: count, dtype: int64
학습 정답 레이블 label
1    30
Name: count, dtype: int64
학습 정답 레이블 label
0    50
1    40
2    30
Name: count, dtype: int64
학습 정답 레이블 label
2    20
1    10
Name: count, dtype: int64
학습 정답 레이블 label
0    50
1    50
2    20
Name: count, dtype: int64
학습 정답 레이블 label
2    30
Name: count, dtype: int64


In [59]:
skf_iris = StratifiedKFold(n_splits=3)
n_iter=0

for train_idx, test_idx in skf_iris.split(iris_df,iris_df['label']):
    n_iter +=1
    label_train = iris_df['label'].iloc[train_idx]
    label_test = iris_df['label'].iloc[test_idx]
    print('학습 정답 레이블', label_train.value_counts())
    print('학습 정답 레이블', label_test.value_counts())


학습 정답 레이블 label
2    34
0    33
1    33
Name: count, dtype: int64
학습 정답 레이블 label
0    17
1    17
2    16
Name: count, dtype: int64
학습 정답 레이블 label
1    34
0    33
2    33
Name: count, dtype: int64
학습 정답 레이블 label
0    17
2    17
1    16
Name: count, dtype: int64
학습 정답 레이블 label
0    34
1    33
2    33
Name: count, dtype: int64
학습 정답 레이블 label
1    17
2    17
0    16
Name: count, dtype: int64


- y값의 희소성으로 발생할 수 있는 문제가
- 공모전 같은 것이나 사기탐지, 이상치 탐지 등
- y값이 희소한 경우 대부분의 거래는 정상이지만 비정상이 있을 수 있고 ( 비정상의 경우는 상당 낮은 경우 )
- 클래스가 불균형이 생길 수 있다.
- 대부분 문제들이 예측을 하는 경우에는 모든 상황이 0,1 이진으로 나눌 때 균형잡히기는 어려울 수 있다.

- 이런 부분을 학습할 때 중요한 포인트이니 학습시에 꼭 불균형을 체크해야 한다.

## 노션 권한 부여되면 그때 과제 제출시되는데
## 미리 과제를 알려드리는 것
## 노션 권한 받으면 수업 이후에 제가 기존에 2주차에 했던 과제도 올려주세요!
## 그때 같이 올려주시면 됩니다.

## 복습과제는 필수입니다.

### 필수과제1
- 타이타닉 데이터를 가지고 위의 과적합이 변경되는지 체크해 주세요!
- 타이타닉 데이터 전처리에 대해서 요약하여 꼭 정리한 내용도 보여주셔야 합니다.

### 필수과제2
- 타이타닉은 조금 작은 데이터라 -> 몇 만? 정도 되는 데이터를 드릴 예정 
- 해당 데이터를 전처리하여 위의 과적합이 실제로 해소되는지를 꼭 체크하기!
- 전처리 내용은 필수적으로 다 적어야 하고

### 결론 도출
- 해당 필수과제1과 필수과제2의 결론을 통해서 본인이 얻은 인사이트를 간략하게 정리해 주세요!